[WIP][New element][tensor_trainer] Create tensor_trainer initial element

songgot opened the pull request:
- Add basic function (event handler, finalize, start, stop, change state, property)
- Add property (framework, input, output, inputtype and outputtype)
- Add caps negotiation (transform_caps, fixate_caps and set_caps)
- Add checking pipeline's input tensor info and tensor's set property values
- Add finding framework and sub-plugin
- Add calling invoke function in transfrom function

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

Signed-off-by: hyunil park <hyunil46.park@samsung.com>

myungjoo reviewed (commented):
Add gsttensor_trainer.c build to android jni build as well.

songgot commented:
I will add gsttensor_trainer.c build to android jni build

myungjoo commented:
If you cannot handle this PR because it is too big (too many commits), create another PR with only a few of the commits to get easily reviewed and merged.

songgot commented:
Okay, thank you.

myungjoo commented on gst/nnstreamer/elements/gsttensor_trainer.c:
```diff
@@ -0,0 +1,1182 @@
+      g_param_spec_string ("model-save-path", "Model save path",
+          "File path to save the model", DEFAULT_STR_PROP_VALUE,
+          G_PARAM_READWRITE | GST_PARAM_MUTABLE_READY |
+          G_PARAM_STATIC_STRINGS));
```
Need more explanataions on the "model-config" and "model-save-path" properties. Some frameworks do not have such files separated. Even if there are two files (config and save), what if the config file has info on the save file?

songgot replied:
Thank you, I will update another PR.
I'll update the statement "If the config file contains information about the save file, it will be ignored".

songgot replied:
Fixed, but... If it's not enough, I'll fix it next time.
Thank you.

myungjoo commented on gst/nnstreamer/elements/gsttensor_trainer.c:
```diff
@@ -0,0 +1,1182 @@
+          "Input tensors dimension from inner array, up to 4 dimensions ?",
+          DEFAULT_STR_PROP_VALUE,
+          G_PARAM_READWRITE | GST_PARAM_MUTABLE_READY |
+          G_PARAM_STATIC_STRINGS));
```
input dimensions are optional, right?

songgot replied:
Thank you.
If we receive it as an optional, we need to pass the value to the sub plugin after parsing the caps in tensor_trainer.
I can change it anytime.

myungjoo commented on gst/nnstreamer/elements/gsttensor_trainer.c:
```diff
@@ -0,0 +1,1182 @@
+          G_PARAM_STATIC_STRINGS));
+
+  g_object_class_install_property (gobject_class, PROP_INPUT_LIST,
+      g_param_spec_uint ("input-lists", "Number of input lists",
```
Number of input lists? And it's string?

songgot replied:
Thank you.
It is changed uint64 (https://github.com/nnstreamer/nnstreamer/pull/4006)

myungjoo commented on gst/nnstreamer/elements/gsttensor_trainer.c:
```diff
@@ -0,0 +1,1182 @@
+          G_PARAM_STATIC_STRINGS));
+
+  g_object_class_install_property (gobject_class, PROP_LABEL_LIST,
+      g_param_spec_uint ("label-lists", "Number of label lists",
```
Wierd name and description, too.

songgot replied:
Thank you, I will update another PR.

myungjoo commented on gst/nnstreamer/elements/gsttensor_trainer.c:
```diff
@@ -0,0 +1,1182 @@
+          G_PARAM_STATIC_STRINGS));
+
+  g_object_class_install_property (gobject_class, PROP_TRAIN_SAMPLES,
+      g_param_spec_uint ("train-samples", "Number of train samples",
```
Make the description more concrete.

myungjoo commented on gst/nnstreamer/elements/gsttensor_trainer.c:
```diff
@@ -0,0 +1,1182 @@
+  g_object_class_install_property (gobject_class, PROP_VALID_SAMPLES,
+      g_param_spec_uint ("valid-samples", "Number of valid samples",
+          "Number of valid samples", 0, G_MAXUINT, 1,
+          G_PARAM_READWRITE | GST_PARAM_MUTABLE_READY |
```
The decription is not enough.

songgot replied:
Thank you, I will update another PR.

songgot replied:
Fixed as much as possible in this PR.
If it's not enough, I'll fix it next time.
Thank you.

myungjoo reviewed (approved):
LGTM. To proceed, let's merge this.
But, please do:
1. make the property description more concrete and readable with full information.
2. you may have some properties renamed (I cannot understand namings of a few properties).
3. make sure that input dimensions/types are optional.

Update such parts in later PRs. Please do not make a huge PR like this. It's unreviewable and it only makes your contribution slower. Make a seperate PR for each topic and make multiple PRs simultaneously if you have many things implemented.

myungjoo commented:
Squash before merge this.

songgot commented:
Thank you.
I will squash these PRs.

songgot commented:
Thank you.
I done squash and added more description and renamed for properties.
but, they may not be enough. I will update next time,
and I'll keep in mind that dimensions and types are optional. it will be update another PR.
Thank you.

again4you commented on gst/nnstreamer/elements/gsttensor_trainer.c:
```diff
@@ -195,6 +220,43 @@ gst_tensor_trainer_class_init (GstTensorTrainerClass * klass)
+  g_object_class_install_property (gobject_class, PROP_NUM_LABELS,
+      g_param_spec_uint ("num-labels", "Number of labels",
+          "A label in a tensor can have one or more classes data,"
+          "set how many labes are recevied", 0, NNS_TENSOR_SIZE_LIMIT, 1,
```
Simple typos
* "set how many labes are recevied" -> "set how many **labels** are **received**"

songgot replied:
Thank you.

songgot replied:
fixed, thank you.

again4you commented on gst/nnstreamer/elements/gsttensor_trainer.c:
```diff
@@ -195,6 +220,43 @@ gst_tensor_trainer_class_init (GstTensorTrainerClass * klass)
+
+  g_object_class_install_property (gobject_class, PROP_NUM_TRAINING_SAMPLES,
+      g_param_spec_uint ("num-training-samples", "Number of training samples",
+          "A sample can consist of muliple inputs and labels in tensors of a gstbuffer"
```
"A sample can consist of muliple inputs and labels in tensors of a gstbuffer"
-> "A sample can consist of **multiple** inputs and labels in tensors of a gstbuffer"

songgot replied:
Thank you.

songgot replied:
fixed, thank you.

again4you commented on gst/nnstreamer/elements/gsttensor_trainer.c:
```diff
@@ -195,6 +220,43 @@ gst_tensor_trainer_class_init (GstTensorTrainerClass * klass)
+  g_object_class_install_property (gobject_class, PROP_NUM_VALIDATION_SAMPLES,
+      g_param_spec_uint ("num-validation-samples",
+          "Number of validation samples",
+          "A sample can consist of muliple inputs and labels in tensors of a gstbuffer"
```
"A sample can consist of muliple inputs and labels in tensors of a gstbuffer"
-> "A sample can consist of **muliple** inputs and labels in tensors of a gstbuffer"

songgot replied:
Thank you.

songgot replied:
fixed, thank you.

again4you reviewed (approved):
Except simple typos, the patches are good to me!
